kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/spec/instantiateSpec.kt

```diff
@@ -24,22 +28,39 @@ suspend fun <T : Spec> createAndInitializeSpec(kclass: KClass<T>, registry: Exte
       null -> runCatching {
          val initial: Spec? = null
 
-         val spec = registry
-            .all()
-            .filterIsInstance<ConstructorExtension>()
+         val spec = constructorExtensions(registry, kclass)
             .fold(initial) { spec, ext -> spec ?: ext.instantiate(kclass) } ?: javaReflectNewInstance(kclass)
 
-         registry
-            .all()
-            .filterIsInstance<PostInstantiationExtension>()
+         postInstantiationExtensions(registry, kclass)
             .fold(spec) { acc, ext -> ext.instantiated(acc) }
       }
+
       else -> Result.success(obj)
    }.onSuccess { spec ->
       spec.globalExtensions().forEach { registry.add(it) }
    }
 }
 
+internal fun constructorExtensions(
+   registry: ExtensionRegistry,
+   kclass: KClass<*>
+): List<ConstructorExtension> {
+   return registry.all().filterIsInstance<ConstructorExtension>() +
+      (kclass.annotation<ApplyExtension>()?.extensions
+         ?.filter { it.isSubclassOf(ConstructorExtension::class) }
```

Comment:
Isn't this already covered by `filterIsInstance<ConstructorExtension>()`?

Reply:
Different sets of extensions. The registry comes from projetc configuration